src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs

```diff
@@ -0,0 +1,370 @@
+        /// intercept each new ServicePoint object being added to ServicePointManager.s_ServicePointTable
+        /// and replace its ConnectionGroupList hashtable field.
+        /// </summary>
+        private class ServicePointHashtable : Hashtable
+        {
+            public override object this[object key]
+            {
+                get
+                {
+                    return base[key];
+                }
+                set
+                {
+                    WeakReference weakRef = value as WeakReference;
+                    if (weakRef != null && weakRef.IsAlive)
+                    {
+                        ServicePoint servicePoint = weakRef.Target as ServicePoint;
+                        if (servicePoint != null)
+                        {
+                            // Replace the ConnectionGroup hashtable inside this ServicePoint object,
+                            // which allows us to intercept each new ConnectionGroup object added under
+                            // this ServicePoint.
+                            Hashtable originalTable = s_connectionGroupListField.GetValue(servicePoint) as Hashtable;
+                            ConnectionGroupHashtable newTable = new ConnectionGroupHashtable();
+                            if (originalTable != null)
+                            {
+                                foreach (DictionaryEntry entry in originalTable)
+                                {
+                                    newTable[entry.Key] = entry.Value;
+                                }
+                            }
+                            s_connectionGroupListField.SetValue(servicePoint, newTable);
+                        }
+                    }
+
+                    base[key] = value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Helper class used for ServicePoint.m_ConnectiopGroupList. The goal here is to
+        /// intercept each new ConnectionGroup object being added to ServicePoint.m_ConnectionGroupList
+        /// and replace its m_ConnectionList arraylist field.
+        /// </summary>
+        private class ConnectionGroupHashtable : Hashtable
+        {
+            public override object this[object key]
+            {
+                get
+                {
+                    return base[key];
+                }
+                set
+                {
+                    if (s_connectionGroupType.IsInstanceOfType(value))
+                    {
+                        // Replace the Connection arraylist inside this ConnectionGroup object,
+                        // which allows us to intercept each new Connection object added under
+                        // this ConnectionGroup.
+                        ArrayList originalArrayList = s_connectionListField.GetValue(value) as ArrayList;
+                        ConnectionArrayList newArrayList = new ConnectionArrayList();
```

Comment:
I am adding a hash table wrapper and an array list wrapper. The idea is that these wrapper are better with race conditions because they reference the same table/list as the underlying data structures. That change should also remove the need for copying.